### DISCUSSION TITLE
Being able to quickly choose the completion item with Alt+Number

### DISCUSSION THREAD

**Discussion by Matthtica** [hrsh7th/nvim-cmp#1482] [Ideas]
I kind of wishing to quickly select the completion item via Alt+Number key without having to scroll through the list with Ctrl+n or any key you use to choose through the completion list
![0xd6wqg9mpma1](https://user-images.githubusercontent.com/74300675/224099971-b09ce28a-fddc-4c5c-8e1a-9f35942bb7e6.png)

**Comment by ghost** [comment]
So [leap](https://github.com/ggandor/leap.nvim) but for the drop down menu. If implementing it won't be too much of a hassle, I can see its use. Seconded.

**Reply by Matthtica** [reply]
The screenshot I show, is from Emacs.
When you press (Alt + 3) for example, it will be auto completed "std::ostream". Kind of like go through one by one and when you reach ostream press enter. But with Alt+Num Bang! One shot.

**Reply by Matthtica** [reply]
It would also be a nice idea to being able to select more than 10 with Alt + Some easy key may be.

**Comment by fitrh** [comment]
You can use `count` opts for `select_{next,prev}_item`, so something like this

```lua
local keys = {}

for i = 1, 10, 1 do
  local key = table.concat({ "<M-", (i < 10 and i or 0), ">" })
  keys[key] = function(fallback)
    if cmp.visible() then
      return cmp.select_next_item({ count = i - 1 })
    end

    fallback()
  end
end

cmp.setup({ mapping = keys })
```

For index label, I believe it can be achieved by using `formatting.format`

**Reply by Matthtica** [reply]
it works, but i am not lua programmer. I usually copy paste for my neovim config. So, i can't write the formatting index label yet.

**Reply by benlubas** [reply]
@fitrh 
> For index label, I believe it can be achieved by using formatting.format

Could you elaborate here? I don't see where in `cmp.Entry` there's information about the order items definitively show up in.

**Comment by benlubas** [comment]
Made a [fork](https://github.com/benlubas/nvim-cmp) with some changes to emulate this a little better (adding numbers, and creating a better function for alt + number key selecting something. This is super hacky and I haven't tested it so it might break horribly. 

No highlighting (yet? idk if I'll add that) and I put the numbers on the right b/c I prefer them there. Easy to change tho if you fork it 
![image](https://user-images.githubusercontent.com/56943754/224890524-43d5cf88-b3c9-4208-a86f-834918ac39df.png)

[dots](https://github.com/benlubas/.dotfiles/blob/87305e67ef62b04b0fc827183dcf06d8b103022e/nvim/lua/plugins/autocomplete.lua#L41-L54) for how to bind keys 

After using this for a day, it's definitely buggy. Sometimes with large menus that are inverted the indexing doesn't work, I'll look into this a little more after work today.

**Reply by benlubas** [reply]
Okay, inverted menus and scrolled menus seem to be working after the most recent changes. If there's interest I can open a PR in a few days when I'm more confident that it's working as intended

**Comment by benlubas** [comment]
#1491

**Comment by tmillr** [comment]
Can't this already be done via the current api and a custom key-mapping?

**Reply by benlubas** [reply]
Unless that api has expanded since this post was last active, no. I don't think you can do it.

Could you explain what you would plan to do with the current api? Maybe I'm missing something.

**Reply by tmillr** [reply]
`:h cmp-function`

```vim
*cmp.visible* ()
  Return a boolean showing whether the completion menu is visible or not.

*cmp.get_entries* ()
  Return all current entries.

*cmp.confirm* (option: cmp.ConfirmOption, callback: function)
  Accepts the currently selected completion item.

*cmp.select_next_item* (option: { behavior = cmp.SelectBehavior, count = 1 })
  Select the next item. Set count with large number to select pagedown.

*cmp.event:on* (%EVENT_NAME%, callback)
  Subscribe to nvim-cmp's event. Events are listed below.

  - `complete_done`: emit after current completion is done.
  - `confirm_done`: emit after confirmation is done.
  - `menu_opened`: emit after opening a new completion menu. Called with a table holding a key
    named `window`, pointing to the completion menu implementation.
  - `menu_closed`: emit after completion menu is closed. Called with a table holding a key
    named `window`, pointing to the completion menu implementation.
```

Just setup the mapping on `menu_opened` event.

> Called with a table holding a key named `window`, pointing to the completion menu implementation.

Visually numbering the entries on screen might require more work though (and I'm not sure if it's even possible, I'm not too familiar with nvim-cmp). You'd also need a way to know which entries are visible, and which aren't (I think). Maybe I'll try to solve this when I've got the time, but it seems like it should be possible one way or the other.
